### PR TITLE
Merging ICRA 2022 challenge updates to master

### PR DESCRIPTION
**Pull request by joydeep-b**
I would like to first create a PR, and then identify what changes need to be reverted, etc. before merging into master. There were a fair number of bug fixes and performance improvements that I'd like to keep.

### CODE REVIEW

File: config/navigation.lua [hunk 1]
```diff
@@ -6,23 +6,23 @@ NavigationParameters = {
   laser_topic = "/velodyne_2dscan";
   odom_topic = "/jackal_velocity_controller/odom";
   localization_topic = "localization";
-  image_topic = "/camera/rgb/image_raw/compressed";
```
Comment (joydeep-b) [inline config/navigation.lua:9]: @elvout @rmeno12 This used to be for VRL-PAP. Should we revert this deletion, or would you prefer to add this back in from your branch once you are done?

File: config/navigation.lua [hunk 2]
```diff
@@ -6,23 +6,23 @@ NavigationParameters = {
...
   init_topic = "initialpose";
   enable_topic = "autonomy_arbiter/enabled";
+  image_topic = "/camera/rgb/image_raw/compressed";
   laser_loc = {
-    x = 0.065;
```
Comment (joydeep-b) [inline config/navigation.lua:13]: Revert.

File: config/navigation.lua [hunk 3]
```diff
@@ -6,23 +6,23 @@ NavigationParameters = {
...
+    x = 0.12;
     y = 0;
   };
   dt = 0.025;
   max_linear_accel = 0.5;
   max_linear_decel = 0.5;
-  max_linear_speed = 0.5;
```
Comment (joydeep-b) [inline config/navigation.lua:19]: Revert all motion params.

File: config/navigation.lua [hunk 4]
```diff
@@ -6,23 +6,23 @@ NavigationParameters = {
...
-  max_angular_accel = 0.5;
-  max_angular_decel = 0.5;
-  max_angular_speed = 1.0;
-  carrot_dist = 10;
+  max_linear_speed = 2.0;
+  max_angular_accel = 5;
+  max_angular_decel = 5;
+  max_angular_speed = 5.0;
+  carrot_dist = 1;
```
Comment (joydeep-b) [inline config/navigation.lua:23]: Let's test this on the real robot.

File: config/navigation.lua [hunk 5]
```diff
@@ -6,23 +6,23 @@ NavigationParameters = {
...
   system_latency = 0.24;
-  obstacle_margin = 0.15;
+  obstacle_margin = 0.05;
```
Comment (joydeep-b) [inline config/navigation.lua:25]: Revert.

File: config/navigation.lua [hunk 6]
```diff
@@ -34,7 +34,8 @@ NavigationParameters = {
   target_dist_tolerance = 0.1;
   target_vel_tolerance = 0.1;
   target_angle_tolerance = 0.05;
-  local_fov = deg2rad(90);
+
+  local_fov = deg2rad(60);
```
Comment (joydeep-b) [inline config/navigation.lua:38]: Let's test this on the real robot.

File: config/ut_jackal.lua [hunk 7]
```diff
@@ -0,0 +1,49 @@
+function deg2rad(deg)
+  return deg * (math.pi / 180)
+end
+
+NavigationParameters = {
```
Comment (joydeep-b) [inline config/ut_jackal.lua:5]: Let's test this on the real robot.

File: config/ut_jackal.lua [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  can_traverse_stairs = false;
+  use_map_speed = true;
+  target_dist_tolerance = 0.1;
+  target_vel_tolerance = 0.1;
+  target_angle_tolerance = 0.05;
+
+  local_fov = deg2rad(60);
+  use_kinect = true;
+  camera_calibration_path = "config/camera_calibration_kinect.yaml";
+  model_path = "../preference_learning_models/jit_cost_model_outdoor_6dim.pt";
```
Comment (joydeep-b) [inline config/ut_jackal.lua:41]: @elvout @rmeno12 Also for VRL-PAP, should we revert?

File: src/navigation/ackermann_motion_primitives.cc [hunk 9]
```diff
@@ -56,10 +57,22 @@ AckermannSampler::AckermannSampler() {
 
 void AckermannSampler::SetMaxPathLength(ConstantCurvatureArc* path_ptr) {
   ConstantCurvatureArc& path = *path_ptr;
+  const float stopping_dist =
+      Sq(vel.x()) / (2.0 * nav_params.linear_limits.max_deceleration);
+  if (!CONFIG_clip_cpoa) {
```
Comment (joydeep-b) [inline src/navigation/ackermann_motion_primitives.cc:62]: This logic seems reversed - shouldn't this be `if (CONFIG_clip_cpoa)`? Investigate.

File: src/navigation/ackermann_motion_primitives.cc [hunk 10]
```diff
@@ -111,51 +120,64 @@ vector<shared_ptr<PathRolloutBase>> AckermannSampler::GetSamples(int n) {
       auto sample = new ConstantCurvatureArc(c);
       SetMaxPathLength(sample);
       CheckObstacles(sample);
-      sample->angular_length = fabs(sample->length * c);
       samples.push_back(shared_ptr<PathRolloutBase>(sample));
     }
-  } else {
+  } else if(false){
```
Comment (joydeep-b) [inline src/navigation/ackermann_motion_primitives.cc:125]: Looks like the sampler changed. If this older procedure is inferior, let's delete it.

File: src/navigation/ackermann_motion_primitives.cc [hunk 11]
```diff
@@ -111,51 +120,64 @@ vector<shared_ptr<PathRolloutBase>> AckermannSampler::GetSamples(int n) {
...
     const float dc = (2.0f * CONFIG_max_curvature) / static_cast<float>(n - 1);
     for (float c = -CONFIG_max_curvature; c <= CONFIG_max_curvature; c+= dc) {
       auto sample = new ConstantCurvatureArc(c);
       SetMaxPathLength(sample);
       CheckObstacles(sample);
-      sample->angular_length = fabs(sample->length * c);
+      samples.push_back(shared_ptr<PathRolloutBase>(sample));
```
Comment (joydeep-b) [inline src/navigation/ackermann_motion_primitives.cc:131]: Test on real robot.

File: src/navigation/ackermann_motion_primitives.cc [hunk 12]
```diff
@@ -111,51 +120,64 @@ vector<shared_ptr<PathRolloutBase>> AckermannSampler::GetSamples(int n) {
...
-      vel.squaredNorm() / (2.0 * nav_params.linear_limits.max_deceleration);
-    if (path.fpl < stopping_dist) {
+    path.length = max(0.0f, path.length);
+    if (path.length < stopping_dist) {
       path.length = 0;
     }
-    
     return;
   }
+  // return;
```
Comment (joydeep-b) [inline src/navigation/ackermann_motion_primitives.cc:180]: Delete.

File: src/navigation/navigation_main.cc [hunk 13]
```diff
@@ -112,8 +113,8 @@ CONFIG_STRING(enable_topic, "NavigationParameters.enable_topic");
 CONFIG_FLOAT(laser_loc_x, "NavigationParameters.laser_loc.x");
 CONFIG_FLOAT(laser_loc_y, "NavigationParameters.laser_loc.y");
 
-DEFINE_string(map, "UT_Campus", "Name of navigation map file");
-DEFINE_string(twist_drive_topic, "navigation/cmd_vel", "Drive Command Topic");
+DEFINE_string(map, "EmptyMap", "Name of navigation map file");
+DEFINE_string(twist_drive_topic, "/cmd_vel", "Drive Command Topic");
```
Comment (joydeep-b) [inline src/navigation/navigation_main.cc:117]: Revert.

File: src/navigation/navigation_main.cc [hunk 14]
```diff
@@ -264,16 +266,64 @@ void SignalHandler(int) {
   run_ = false;
 }
 
+
```
Comment (joydeep-b) [inline src/navigation/navigation_main.cc:269]: Delete.

File: src/navigation/navigation_main.cc [hunk 15]
```diff
@@ -264,16 +266,64 @@ void SignalHandler(int) {
...
     map = msg.map;
     navigation_.UpdateMap(navigation::GetMapPath(FLAGS_maps_dir, msg.map));
   }
+  if (executed_trajectory_.size() > kMaxTrajectoryLength) {
+    executed_trajectory_.erase(executed_trajectory_.begin());
+  }
+  executed_trajectory_.push_back(Vector2f(msg.pose.x, msg.pose.y));
+}
+
+void LocalizationSpoofedCallback(const nav_msgs::Odometry& msg) {
```
Comment (joydeep-b) [inline src/navigation/navigation_main.cc:309]: We can probably keep this, but we should document it and give it a better name.

File: src/navigation/navigation_main.cc [hunk 16]
```diff
@@ -422,20 +472,20 @@ void DrawTarget() {
   visualization::DrawCross(target, 0.2, 0x10E000, msg_copy);
   visualization::DrawArc(
       Vector2f(0, 0), carrot_dist, -M_PI, M_PI, 0xE0E0E0, local_viz_msg_);
-  viz_pub_.publish(msg_copy);
+  // viz_pub_.publish(msg_copy);
```
Comment (joydeep-b) [inline src/navigation/navigation_main.cc:475]: Not sure why this is commented out.

File: src/navigation/navigation_main.cc [hunk 17]
```diff
@@ -844,48 +903,48 @@ int main(int argc, char** argv) {
       n.subscribe("halt_robot", 1, &HaltCallback);
   ros::Subscriber override_sub =
       n.subscribe("nav_override", 1, &OverrideCallback);
-
+  //ros::Subscriber path_sub = n.subscribe("/move_base/TrajectoryPlannerROS/global_plan",1,&PathCallback);
```
Comment (joydeep-b) [inline src/navigation/navigation_main.cc:906]: Delete?

File: src/navigation/navigation_main.cc [hunk 18]
```diff
@@ -844,48 +903,48 @@ int main(int argc, char** argv) {
...
+    RateChecker::Invocation rc_iteration(&rate_checker);
     visualization::ClearVisualizationMsg(local_viz_msg_);
     visualization::ClearVisualizationMsg(global_viz_msg_);
     ros::spinOnce();
 
     // Run Navigation to get commands
     Vector2f cmd_vel(0, 0);
     float cmd_angle_vel(0);
-    bool nav_succeeded = navigation_.Run(ros::Time::now().toSec(), cmd_vel, cmd_angle_vel);
+    navigation_.Run(ros::Time::now().toSec(), cmd_vel, cmd_angle_vel);
```
Comment (joydeep-b) [inline src/navigation/navigation_main.cc:925]: We should still get the success flag and think harder about what to do when it fails. We can't just not publish visualizations if it fails. Need a better way to handle failures.

File: src/visualization/visualization.cc [hunk 19]
```diff
@@ -136,7 +136,7 @@ void DrawPathOption(const float curvature,
       DrawLine(
           Vector2f(0, clearance), Vector2f(distance, clearance), color, msg);
       DrawLine(
-          Vector2f(0, clearance), Vector2f(distance, clearance), color, msg);
+          Vector2f(0, -clearance), Vector2f(distance, -clearance), color, msg);
```
Comment (joydeep-b) [inline src/visualization/visualization.cc:139]: Caught a bug!

File: CMakeLists.txt [hunk 20]
```diff
@@ -44,59 +44,25 @@ SET(LIBRARY_OUTPUT_PATH ${PROJECT_SOURCE_DIR}/lib)
...
-                        src/navigation/linear_evaluator.cc
-                        src/navigation/image_tiler.cc
-                        src/navigation/deep_cost_map_evaluator.cc
-                        src/navigation/image_based_evaluator.cc
-                        )
-TARGET_LINK_LIBRARIES(navigation 
-    shared_library ${libs} ${TORCH_LIBRARIES} ${OpenCV_LIBS})
-
-
-ADD_EXECUTABLE(navigation_tests
```
Comment (joydeep-b) [inline CMakeLists.txt:78]: We should restore the navigation tests.

**Comment by OAHC2022** [comment]
outdated. will close

**Pull request closed without merging**